Feat/1544 add nginx configuration increase security performance (#2173)
raft-tech/TANF-app: tdrs-frontend/nginx/README.md

# Nginx Documentation

## 1. Installation

The Nginx uses a master process and mutiple workers to efficiently manage requests. The master process reads and evaluates config file and manages the workers. 

In TDRS, all config and associated files with it are placed in ```/tdrs-frontend/nginx``` folder. There are various differences between the local and deployed versions. Due to the fact that cloud.gov uses buildpacks and local version uses containerized docker instance, there are two separate directories for local and deployed versions.

## 2. Architecture

In general for TDRS, the *Nginx* handles incoming HTTP requests to both frontend and backend. The frontend requests are handled directly while the backend requests are forwarded to the backend server.

> ![Cloud.gov Architecture](./src/arch1.jpg)

Reviewer: We might be able to keep a single diagram and just update this one:

https://github.com/raft-tech/TANF-app/blob/develop/docs/Security-Compliance/boundary-diagram.md

> *Nginx* serves the frontend built files placed in *```/usr/share/nginx/html```* which ensures not  only the static files are served efficiently, the security headers area added as defined in the conf file.

Reviewer: ```suggestion
*Nginx* serves the frontend built files placed in *```/usr/share/nginx/html```* which ensures not only the static files are served efficiently, the security headers area added as defined in the conf file.
```

- HTTP requests are first handled by Cloud.gov (.app.cloud.gov subnet) and then routed to frontend instance
- Nginx serves frontend requests directly and routes backend api requests to backend server.

> - The backend requests are distinguished by their specific URI (e.g: addresses starting with ```/v1```) usinf location directive (see below).

Reviewer: ```suggestion
- The backend requests are distinguished by their specific URI (e.g: addresses starting with ```/v1```) using location directive (see below).
```

        proxy_connect_timeout 60;
        proxy_send_timeout    60;
        proxy_buffer_size     4k;
        proxy_temp_file_write_size 64k;
}
```

The Nginx configuration files are located in: ```*'/frontend/nginx/'```*

## Backend

### Nginx

The frontend then sends processing requests to the *backend* Django server, which are handled by *[gunicorn](https://gunicorn.org)*. See section below [Gunicorn](###Gunicorn).

### Gunicorn

Gunicorn is WSGI HTTP server based on python and uses **worker [Worker Processes](https://docs.gunicorn.org/en/stable/design.html#choosing-a-worker-type)**. The number of workers is relative to server request load. 

With having workers responding to requests in Gunicorn, it is importnat to have Nginx in front to handle requests first, otherwise DDOS attacks would consume the server (See Nginx config).

Gunicorn by default uses **[Sync worker](https://docs.gunicorn.org/en/latest/design.html#sync-workers)**. This is the most basic worker with some implications. Sync does not support persistent connections, which is the case in our application.
```python
# The number of worker processes for handling requests
workers = 2
```

> There are two config files: one for development and one for deployed application. Gunicorn is started from ```gunicorn_start.sh``` which starts either version of config files based on envirenment.

Reviewer: ```suggestion
There are two config files: one for development and one for deployed applications. Gunicorn is started from ```gunicorn_start.sh``` which starts either version of config files based on environment.
```

## Security

### Whitelist IPs

> A list of IP addresses has been added to ```ip_whitelist.conf```. This means any request from an ip address not in the subnets included in this file will be rejected. This list is created manualy and needs to be maintained to whitelist and include user IP subnets.

Reviewer: ```suggestion
A list of IP addresses has been added to ```ip_whitelist.conf```. This means any request from an ip address not in the subnets included in this file will be rejected. This list is created manually and needs to be maintained to whitelist and include user IP subnets.
```

It might be good context/history to explain why we implemented this at all and why here specifically.
